modules/microsite/docs/scala/akka-http/generating-a-server.md

```diff
@@ -57,20 +57,20 @@ trait UserHandler[-E] {
 }
 
 object UserResource {
-  def routes[E](handler: UserHandler, customExtract: String => Directive1[E])(implicit mat: akka.stream.Materializer): Route = {
+  def routes[E](handler: UserHandler, customExtract: String => Directive[E])(implicit mat: akka.stream.Materializer): Route = {
     // ...
   }
 }
 ```
 
-You may now provide a function that accepts the `operationId` of the route, and returns a `Directive1[E]`, where, `E` is an arbitrary type that will be passed through to your handlers. The `Directive1[E]` will be injected into the generated routes _after_ the path and method `Directive`s.
+You may now provide a function that accepts the `operationId` of the route, and returns a `Directive[E]`, where, `E` is a tuple with arbitrary types that will be passed through to your handlers. The `Directive[E]` will be injected into the generated routes _after_ the path and method `Directive`s. Note that extracting a single value still requires `E` to be a `Tuple1`.
 
 For example, to extract an `X-User-Id` header value from an incoming request, your code might look like this:
 
 ```scala
-class UserApi extends UserHandler[String] {
-  override def createUser(respond: UserResource.CreateUserResponse.type)(body: definitions.User)(userIdHeader: String): scala.concurrent.Future[UserResource.CreateUserResponse] = {
-    println(s"The supplied X-User-Id header is: $userIdHeader")
+class UserApi extends UserHandler[Tuple1[String]] {
+  override def createUser(respond: UserResource.CreateUserResponse.type)(body: definitions.User)(extracted: Tuple1[String]): scala.concurrent.Future[UserResource.CreateUserResponse] = {
+    println(s"The supplied X-User-Id header is: ${extracted._1}")
```

Comment:
this is really unpleasant, thanks for pointing it out.

Reply:
So, despite my best efforts, the only thing I could come up with was:
```scala
object pkg {
  sealed class DirectiveAdapter[A](val value: Directive1[A])

  sealed trait DirectiveAdapterLowPriorityImplicits {
    implicit def fromDirective[A]: Directive[A] => DirectiveAdapter[A] = d => new DirectiveAdapter(d.tmap(Tuple1(_)))
  }
  object DirectiveAdapter extends DirectiveAdapterLowPriorityImplicits {
    implicit def fromDirective1[A]: Directive1[A] => DirectiveAdapter[A] = d => new DirectiveAdapter(d)
  }
  def routes[A](extract: DirectiveAdapter[A]): Directive1[A] = extract.value
}
```
which permits
```scala
pkg.routes[Unit](Directives.get)
pkg.routes[String](Directives.headerValueByName("foo"))
val foobar = (Directives.headerValueByName("foo") & Directives.headerValueByName("bar"))
pkg.routes[(String, String)](foobar)
```

*This is bad because* it does implicit conversion of values, and is broadly undiscoverable.

```scala
object pkg {
  import scala.language.higherKinds
  sealed trait DirectiveAdapter[F[_]] { def convert[A]: F[A] => Directive1[A] }
  sealed trait DirectiveAdapterLowPriorityImplicits {
    implicit object DirectiveAdapterN extends DirectiveAdapter[Directive] { def convert[A] = _.tmap(Tuple1(_)) }
  }
  object DirectiveAdapter extends DirectiveAdapterLowPriorityImplicits {
    implicit object DirectiveAdapter1 extends DirectiveAdapter[Directive1] { def convert[A] = identity _ }
  }
  def routes[F[_], A](extract: F[A])(implicit ev: DirectiveAdapter[F]): Directive1[A] = ev.convert(extract)
}
```
is _technically_ better, though `def routes[F[_], A]` is 👎 from a complexity perspective.

this is quite the puzzler.

#### To clarify

I'm now very much against just introducing `Tuple1` wholesale because propagating `._1` everywhere is really a terrible solution. The amplification of bad code is expressly against the core philosophy of why guardrail exists, we gotta defend this every chance we can get, since the internals of guardrail is much easier to change than the interfaces consumed by users.

Reply:
Yeah as soon as I saw those tuple shenanigans I began to dislike that solution a bit more too.

Another thought: what if we actually had the user configure how many params to extract? If they want to extract 2 params, then we could generate signatures like:

```
def routes[E1, E2](handler: UserHandler, customExtract: String => Directive[(E1, E2)])

trait UserHandler[-E1, -E2] {
  def createUser(respond: UserResource.CreateUserResponse.type)(body: definitions.User)(extracted1: E1, extracted2: E2): scala.concurrent.Future[UserResource.CreateUserResponse]
  // ...
}
```

I didn't actually try this, but seems like it would work. I'm now headed into the woods for a week so can't work on it for a bit, but what do you think?

Alternately, maybe the original `Directive1` solution isn't too bad. I do think most people will be extracting a single thing, and we could document how to use `tmap` to compose multiple extractions.

Reply:
IMO the `[E1, E2]` strategy:
- requires configuring guardrail to explicitly align with individual use cases, precluding multiple invocations of `Handler` with different parameters in the same compilation unit
- ends up seizing all type parameters to `routes` for itself for complex services

To this end, the `[F[_], A]` approach is preferable, though the lack of symmetry in `routes[Directive1, String]` and `routes[Directive, (String, String)]` is not excellent.

While it's still not ideal, I think leaving `Directive1` and moving forward with `tmap(Tuple1(_))` for the time being, then adding this complexity down the line if there's significant motivation to do so.

Reply:
I'll revert back to your previous strategy, merge, and release.